Rampis/mt7620

hungaborhorvath opened the pull request:
This pull request is based on #764. 

Update ramips/mt7620 profile to Openwrt 18.06.1

Merged usb_mt7620 and usb_mt7620_mt76, since there was virtually no difference left in openwrt 18.06.1. Added mmc support, since some of the routers have micro sd card reader inside, and the packages do not take up much space.

The config file was generated by 'make custom'. From the menuconfig the following were chosen:

Target System:
    Mediatek Ralink MIPS
Global Build Settings ->
    Select "Compile with support for patented functionality"
Administration -> Gargoyle Web Interface ->
    Select gargoyle, gargoyle-i18n, plugin-gargoyle-i18n-English-EN,
           plugin-gargoyle-openvpn, plugin-gargoyle-usb-printer,
           plugin-gargoyle-usb-storage, and leave the rest as "m"
Kernel modules -> Other modules ->
    Select "kmod-sdhci-mt7620"
Network ->
    plugin-gargoyle-ddns-cloudflare as "m"

Then save the config (twice, because for the first time several commented
ar71xx modules are included, as well), then quit from menuconfig and CTRL+C.

Target ramips needed to be patched so build would not fail if the generated image is too big. Profile usb_mt7620 has all the profiles now that compile without problems. Dropping the micro sd support does not add any new devices into the list.

Created mt7620_mini profile, which only contains the gargoyle web interface and the internationalization package. All the rest from above are included as modules. Three devices do not compile, even with internationalization dropped:
    -ex2700
    -vonets_var11n-300
    -wt3020-4M

Replaced dependency of msmtp on ca-bundle to ca-certificates-combined.

lantis1008 commented:
Sounds good. 
Did you see that I sent you a frame work for the first bits that need patching on the forum?

hungaborhorvath commented:
Yes, thanks. I replied there to your post a couple minutes ago.

lantis1008 commented:
Seems you’ve got the hang of the patches now. It’s a shame that the default behaviour crashes the entire build. 
When I recently raised this with some people they couldn’t understand why you wouldn’t want it to crash.

hungaborhorvath commented:
Yeah, I understand.

Actually, I am a bit stuck. The next crash happens when ````zyimage```` is called for the factory image for ````kn_rc```` . Now, I put a ````; \```` at the end of line 17 of ````ramips-src/target/linux/ramips/image/mt7620.mk````, and an ````echo ""```` in the next ilne but somehow the build still crashes. Maybe I should put a space before the ````;```` because that line ended with ````$@````? But that really shouldn't matter....

lantis1008 commented:
Is it zyimage failing or is it check-size? 
If it's zyimage i think you'll want to do something like the following:
```
define Build/zyimage
         if [ -e "$1" ] ; then \
                  $(STAGING_DIR_HOST)/bin/zyimage $(1) $@ \
         fi;
endef
```

hungaborhorvath commented:
Ok, so actually that was not the only error, there was also one in ````pad-to````. However, the if-then version did not work, no matter which way I tried. In the end, I just put a semicolon at the end of the respective lines, and added an extra ````echo ""```` to make sure to return 0 exit code. 

In fact, I am not sure why we do all the if-then magic. Openwrt makes sure to delete the too large files, so all we really need is to have exit code 0 everywhere for the run to continue. That can be achieved if we concatenate the relevant lines by adding ````;\```` to the end of each line, and finish with a command that returns 0 exit code always (````echo ""````). This should be enough. 

Am I missing something?

lantis1008 commented:
Some of the programs that are called fail immediately if the input file doesn’t exist and seems to end the build. 
Covering the non existent files with an it seems necessary in at least some cases.

hungaborhorvath commented:
Ok, I think I have it now. 30 devices out of 75 compile with openvpn, usb and micro sd support. Dropping the micro sd support does not help in getting more devices. I created an mt7620_mini profile, which essentially only contains the gargoyle web interface with English internationalization. I guess this way everyone can decide what is more important to them: usb or openvpn, and then they can install whichever they like and have the space for. 3 devices do not compile even with this bare minimum, so they cannot run gargoyle. 

Some testing is still in order, before I remove the Do not merge flag. 

Also, I would welcome any suggestions on how to clean up the history, etc. E.g. if you simply want one commit instead of many (on top of #764 of course), I don't mind squashing them and rewriting the log. 

I am not sure about the if-then situation mentioned above. If it is needed, I can try to make all patch changes always (or never) using if-then, but I am not sure if that is necessary. Right now the patches seem to work for these two targets..

lantis1008 commented:
The commit history seems ok to me. I wouldn’t worry about that. 

If the patches work then don’t worry. We can adjust them later if something breaks. 
Good job. 👍

hungaborhorvath commented:
Ok, I rewrote the first post of the PR, so the text could be used as a log entry when merged. 

Removed the "Do not merge" flag, both profiles seemed to compile ok after ````make custom````.

lantis1008 commented on targets/ramips/profiles/usb_mt7620/config:
```diff
@@ -2270,96 +3529,318 @@ CONFIG_PACKAGE_webmon-gargoyle=y
+# CONFIG_PACKAGE_pagekitec is not set
+# CONFIG_PACKAGE_pen is not set
+# CONFIG_PACKAGE_pingcheck is not set
+# CONFIG_PACKAGE_plugin-gargoyle-ddns-cloudflare is not set
```
Please include this as a module. It will require curl/libcurl as well (which it should select automatically). It is in a different section of the configuration menu to the other Gargoyle modules.

hungaborhorvath replied:
Hm, ok, I completely missed this one, I will look through the menu.

hungaborhorvath replied:
This should be resolved now.

lantis1008 commented on targets/ramips/profiles/usb_mt7620/config:
```diff
@@ -2368,41 +3849,77 @@ CONFIG_PACKAGE_wwan=y
 CONFIG_PACKAGE_disktype=y
-CONFIG_PACKAGE_dosfsck=y
-# CONFIG_PACKAGE_dosfslabel is not set
+# CONFIG_PACKAGE_dosfstools is not set
```
I thought this one was required by Gargoyle. Not sure.

hungaborhorvath replied:
Surely not if this is not in the config file. The config files are generated using menuconfig in the way I explained in the log for commit e62cb58 If it were a dependency, the menuconfig would have added it.

Anyway, do you want me to add it as dependency for gargoyle, or shall we just leave it as is?

hungaborhorvath commented on targets/ramips/profiles/usb_mt7620/config:
```diff
@@ -1185,38 +1392,38 @@ CONFIG_BUSYBOX_DEFAULT_FEATURE_SH_NOFORK=y
-# CONFIG_BUSYBOX_DEFAULT_KLOGD is not set
-# CONFIG_BUSYBOX_DEFAULT_FEATURE_KLOGD_KLOGCTL is not set
-CONFIG_BUSYBOX_DEFAULT_LOGGER=y
+CONFIG_PACKAGE_ca-bundle=m
```
Without this requirement the package msmtp threw an error to me and didn't build, and you fixed it in ````package/msmtp/Makefile```` by adding it as a dependency: https://www.gargoyle-router.com/phpbb/viewtopic.php?f=8&t=11430&sid=1573575bf75f468070ebd5d68ceab143&start=20#p50216

Or are you saying that I should just give ca-certificates-combined as dependency of msmtp, instead of ca-bundle?

hungaborhorvath replied:
This should be resolved now.

hungaborhorvath commented on targets/ramips/profiles/usb_mt7620/config:
```diff
@@ -470,39 +577,51 @@ CONFIG_BUSYBOX_DEFAULT_FEATURE_DU_DEFAULT_BLOCKSIZE_1K=y
 CONFIG_BUSYBOX_DEFAULT_FEATURE_LS_COLOR=y
 CONFIG_BUSYBOX_DEFAULT_FEATURE_LS_COLOR_IS_DEFAULT=y
 CONFIG_BUSYBOX_DEFAULT_MD5SUM=y
+# CONFIG_BUSYBOX_DEFAULT_SHA1SUM is not set
```
Is it ok if I give it as a dependency for ````plugin-gargoyle-usb-storage````?

lantis1008 replied:
Yes, but make it a dependency of `share_users` instead. `usb-storage` relies on share_users so the dependency will carry through.

hungaborhorvath replied:
Hm, now that I think about it, I am not sure how to make this as dependency or provided by share-users. This is not a package, but some setting in menuconfig. What should I write into package/share-users/Makefile ?

hungaborhorvath replied:
Ok, I added something which I thought should work based on https://wiki.openwrt.org/doc/devel/dependencies but somehow it does not reflect in the config file. What did I do wrong?

lantis1008 replied:
I’ve asked for some assistance from the Openwrt devs on this one. 
I’m sure it is doable, but I don’t know.

lantis1008 replied:
Making the config file bigger is ok.

We either include sha1sum or rewrite everything that relies on it. 
At the time it was easier for me to add it back.

hungaborhorvath replied:
Hm, I don't like this option. If we set Customize busybox options, then (apart from having a 30k larger config file) it would mark the build tainted for openwrt. 

I think the long term solution should be to get rid of the sha1sum dependency. 

However, for the short term I think it would still be more clean to patch ````/ramips-src/package/utils/busybox/Config-defaults.in```` to have ````BUSYBOX_DEFAULT_SHA1SUM```` default to ````y```` instead of ````n````. And this would then work for all gargoyle builds, not only the ramips/mt7620. 

TL;DR: I would prefer pathcing ````/ramips-src/package/utils/busybox/Config-defaults.in```` over having a 30k bigger config file. Would this be an agreeable solution?

lantis1008 replied:
Sure. But you still need to add it to your current config of course

lantis1008 commented:
Appreciate it. They did get back to me but not with particularly useful information. 
It may be best to remove the dependency for now if it cannot be resolved and just know that sha1sum is a requirement.

hungaborhorvath commented:
What I can do is: do a make custom with ````CONFIG_BUSYBOX_DEFAULT_SHA1SUM```` enabled, and see how much of the config file changes. Then I can add it as a last commit, and it can be easily dropped if not needed, or whatever. This of course doesn't solve similar issues later on, but maybe let's do one step at a time. 

I thought some more about the history: 
I think the cleanest solution would be that commit c345474 is squashed into your commit 41c0535, but that would mean rewriting the history of the commits of your PR. 

That is either you do it in your PR and I rebase this PR (I would prefer it), or I do it here, but then your PR and this one would be out of sync.

lantis1008 commented:
I would drop commit c345474 entirely. 
If choose not to, it will still apply cleanly once my PR is merged I think. 

But I wouldn’t go trying too hard to rewrite history. It can be fixed during merge easily enough by @ericpaulbishop if needed

hungaborhorvath commented:
Ok, I think I resolved the last thing. Anything else I should do?

lantis1008 commented:
No that’s fine. Thank you for your work

lantis1008 commented:
Besides the merge conflict. The only thing missing from this is ramips/default as far as 1.11.0 goes. I’ve been working on this slowly.
